packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+          name,
+          statics
+        );
+        let value =
+          attributePart.strings === undefined
+            ? result.values[partIndex]
+            : result.values;
+        value = attributePart._resolveValue(value, partIndex);
+        const instance = op.useCustomElementInstance
+          ? getLast(renderInfo.customElementInstanceStack)
+          : undefined;
+        if (prefix === '.') {
+          const propertyName = name.substring(1);
+          // Property should be reflected to attribute
+          const reflectedName = reflectedAttributeName(
+            op.tagName,
+            propertyName
+          );
+          // Property should be set to custom element instance
+          const instance = op.useCustomElementInstance
+            ? getLast(renderInfo.customElementInstanceStack)
+            : undefined;
+          if (instance || reflectedName !== undefined) {
+            if (value !== noChange) {
+              if (instance !== undefined) {
+                instance.setProperty(propertyName, value);
+              }
+              if (reflectedName !== undefined) {
+                yield `${reflectedName}="${escapeAttribute(String(value))}"`;
+              }
+            }
+          }
+        } else if (prefix === '@') {
+          // Event binding, do nothing with values
+        } else if (prefix === '?') {
+          // Boolean attribute binding
+          attributeName = attributeName.substring(1);
+          if (value && value !== noChange) {
+            yield attributeName;
+          }
+        } else {
+          if (value !== noChange) {
+            if (instance !== undefined) {
+              instance.setAttribute(attributeName, value as string);
+            } else {
+              yield `${attributeName}="${escapeAttribute(String(value))}"`;
+            }
+          }
+        }
+        partIndex += statics.length - 1;
+        break;
+      }
+      case 'custom-element-open': {
+        const ctor = op.ctor;
+        // Instantiate the element and stream its render() result
+        let instance = undefined;
+        try {
+          const element = new ctor();
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (element as any).tagName = op.tagName;
+          // TODO: Move renderer instantiation into a plugin system
+          if (element instanceof LitElement) {
+            instance = new LitElementRenderer(element);
+          } else {
+            console.error(`No renderer for custom element: ${op.tagName}`);
+          }
+        } catch (e) {
+          console.error('Exception in custom element constructor', e);
+        }
+        renderInfo.customElementInstanceStack.push(instance);
+        break;
+      }
+      case 'custom-element-attributes': {
+        const instance = getLast(renderInfo.customElementInstanceStack);
+        if (instance !== undefined) {
+          if (instance.connectedCallback) {
```

Comment:
Why is this order important (connected before attributes)? It doesn't match spec.

Reply:
`renderAttributes` here is _emit them out to the stream_, not set them to the instance.

We call `renderer.connectedCallback()` so the element can do any connect-time work (e.g. LitElement reflects attributes) before we emit the attributes to the stream.

In `custom-element-open` we set static attributes into the instance and in `attribute-part` we set any attribute/property bindings, after calling `connectedCallback` the state of the attributes on the instance are g2g, so we can emit them.

Reply:
I added some comments to this effect.

Reply:
ok